[JIT] Add cond-add-relu matching pattern to cover in-place ops

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#55458 [JIT] Add cond-add-relu matching pattern to cover in-place ops**

Summary: Previously the cond-add-relu pass blindly turns all in-place add and relu ops into non-mutation version, when those ops are not part of the fusion patten, it can actually hurt the performance as shown in densenet on some platform.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27620415](https://our.internmc.facebook.com/intern/diff/D27620415)

desertfire committed:
[JIT] Add cond-add-relu matching pattern to cover in-place ops

Summary: Previously the cond-add-relu pass blindly turns all in-place add and relu ops into non-mutation version, when those ops are not part of the fusion patten, it can actually hurt the performance as shown in densenet on some platform.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

eellison reviewed (approved):
LGTM - one question, when you match `conv_->add_->relu_` do we guarantee that `add_` is on the `conv` output and not on another input? For instance, rewriting the following is not safe, it would have to be `x.add_(y)` insteaed
```
x = Conv(...)
z = y.add_(x)
return torch.relu(z)
```

desertfire committed:
Update on "[JIT] Add cond-add-relu matching pattern to cover in-place ops"


Summary: Previously the cond-add-relu pass blindly turns all in-place add and relu ops into non-mutation version, when those ops are not part of the fusion patten, it can actually hurt the performance as shown in densenet on some platform.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27620415](https://our.internmc.facebook.com/intern/diff/D27620415)

[ghstack-poisoned]
